src/layout.cpp

```diff
@@ -0,0 +1,476 @@
+  const TreeNode *node = getNode(id);
+  assert(node);
+  return node->succ;
+}
+
+const std::vector<TreeNode::nodeId> &
+Net::getPredecessors(TreeNode::nodeId id) const {
+  const TreeNode *node = getNode(id);
+  assert(node);
+  return node->pred;
+}
+
+const TreeNode *Net::getNode(TreeNode::nodeId id) const {
+  if (id >= 0 && id < nodes.size()) {
+    return nodes.data() + id;
+  }
+  return nullptr;
+}
+
+void Net::transformationGraph() {
+  std::vector<std::pair<TreeNode::nodeId, TreeNode::nodeId>> deletedEdges = {};
+  greedyFAS(nodes, deletedEdges);
+
+  std::vector<int> lensLayer = {};
+  algorithmASAP(nodes, deletedEdges, lensLayer);
+
+  addAllDummyNodes(nodes, lensLayer);
+
+  predReinit(nodes);
+  degReinit(nodes);
+}
+
+void Net::netTreeNodesToNormalizedElements(
+    std::vector<NormalizedElement> &normalizedElements) {
+  float maxNumber = -1, maxLayer = -1;
+  for (int i = 0; i < nodes.size(); i++) {
+    if (nodes[i].layer > maxLayer) {
+      maxLayer = nodes[i].layer;
+    }
+    if (nodes[i].number > maxNumber) {
+      maxNumber = nodes[i].number;
+    }
+  }
+  float nCellSize = -1;
+
+  if (maxLayer > maxNumber) {
+    nCellSize = 1.0 / (maxLayer + 1);
+  } else {
+    nCellSize = 1.0 / (maxNumber + 1);
+  }
+
+  for (int i = 0; i < nodes.size(); i++) {
+    NormalizedElement nElement = {};
+    nElement.id = nodes[i].id;
+
+    if (nodes[i].isDummy) {
+      NormalizedPoint nPoint = {};
+      nPoint.nX = nCellSize * nodes[i].number + (nCellSize / 2) / 2;
+      nPoint.nY = nCellSize * nodes[i].layer + (nCellSize / 4) / 2;
+
+      nElement.nPoint = nPoint;
+      nElement.nH = 0;
+      nElement.nW = 0;
+    } else {
+      NormalizedPoint nPoint = {};
+      nPoint.nX = nCellSize * nodes[i].number;
+      nPoint.nY = nCellSize * nodes[i].layer;
+
+      nElement.nPoint = nPoint;
+      nElement.nH = nCellSize / 4;
+      nElement.nW = nCellSize / 2;
+    }
+
+    normalizedElements.push_back(nElement);
+  }
+
+  int countConnections = 0;
+  for (int i = 0; i < nodes.size(); i++) {
+    for (int j = 0; j < nodes[i].succ.size(); j++) {
+      NormalizedConnection connection = {};
+
+      connection.id = countConnections;
+      connection.startElementId = nodes[i].id;
+      connection.endElementId = nodes[i].succ[j];
+
+      NormalizedPoint nPointStart = {};
+      nPointStart.nX =
+          normalizedElements[i].nPoint.nX + normalizedElements[i].nW / 2;
+      nPointStart.nY =
+          normalizedElements[i].nPoint.nY + normalizedElements[i].nH;
+
+      NormalizedPoint nPointEnd = {};
+      nPointEnd.nX = normalizedElements[nodes[i].succ[j]].nPoint.nX +
+                     normalizedElements[nodes[i].succ[j]].nW / 2;
+      nPointEnd.nY = normalizedElements[nodes[i].succ[j]].nPoint.nY;
+
+      connection.nVertices.push_back(nPointStart);
+      connection.nVertices.push_back(nPointEnd);
+
+      countConnections++;
+
+      normalizedElements[i].connections.push_back(connection);
+    }
+  }
+}
```

Comment:
Split into several procedures with proper names.
Use shorter names when appropriate.
Avoid using arbitrary constants -- for common calculations create named functions.
Avoid using nested indexing, use named references instead.